### DISCUSSION TITLE
How do I prevent leaving a page ?  (kind of "are you sure you don't want to save")

### DISCUSSION THREAD

**Discussion by vindevoy** [zauberzeug/nicegui#1700] [Q&A]
### Question

Hi,

I have a menu linking to several pages.  In one page, when I do things (I'm building a graph) I want to ask the user if he/she is sure to leave the page when he/she clicks another menu (or even the same menu as currently selected).

How can this be done ?

**Comment by natankeddem** [comment]
In general I would utilize `ui.dialog` for that sort of interaction probably. If you provided a minimum example of your existing page I could try to help further.

**Comment by falkoschindler** [comment]
Interesting question, @vindevoy! I guess you mean something like the dialog which appears, e.g., on https://github.dev/zauberzeug/nicegui/tree/main/nicegui when you press Cmd-W (Mac) or Ctrl-W (else).

<img width="372" alt="Screenshot 2023-09-27 at 17 26 27" src="https://github.com/zauberzeug/nicegui/assets/5767091/f786db90-6fbc-4a02-9fd9-7eb674b0608f">

One solution might be to add some JavaScript to the body:

```py
ui.add_body_html('''
<script>
window.addEventListener("beforeunload", function (e) {
    var confirmationMessage = 'Are you sure you want to leave?';
    e.returnValue = confirmationMessage; // Standard for some browsers
    return confirmationMessage; // Others
});
</script>
''')
```

**Reply by vindevoy** [reply]
Yes, this does what it should.  Or at least, what I asked.

But, the "Are you sure you want to leave ?" is not shown.  It simply shows "Leave Site ?  Changes may not be saved." on Edge and Chrome.  

Now remains: how do I get the value of the button they clicked ?  Should I bind it to something ?  And how can I execute a Python function on the "Leave" button.

**Reply by falkoschindler** [reply]
Regarding the specific message you're seeing, this is what ChatGPT explained to me:
> Historically, developers could set custom messages that would be shown in the dialog when the `beforeunload` event was triggered. But some websites abused this to confuse or mislead users, or to create "exit traps" where it was difficult for users to leave the website.
>
> To mitigate these issues, many browsers, including Chrome, decided to either ignore custom messages or provide a standardized message. This means that even if you try to return a custom string from the `beforeunload` event handler, like `'Are you sure you want to leave?'`, Chrome will show its own default message instead.
>
> Therefore, in browsers like Chrome, the purpose of the `return` value is not to set a custom message but to simply signal to the browser that it should display its own confirmation dialog. The actual text shown in the dialog is determined by the browser and cannot be customized.

**Reply by falkoschindler** [reply]
To trigger a Python function when the user wants to leave the page, you can register a custom event like this:
```py
before_unload = ui.button().on('beforeunload', lambda: print('beforeunload')).classes('hidden')
ui.add_body_html(f'''<script>
window.onbeforeunload = function() {{
    getElement({before_unload.id}).$emit('beforeunload');
    return 'Are you sure you want to leave?';
}};
</script>
''')
```
Similar to my solution in https://github.com/zauberzeug/nicegui/discussions/1731#discussioncomment-7190949 this uses a button just to have an element for registering the event.
> In order to register a custom event, we need a Vue element. Here I chose a ui.button, but it could be another one as well. It is hidden anyway. (Somehow ui.element doesn't work... We should think about a better way for registering such "global" events. But for now it's working.)

Note that this event is emitted as soon as the dialog opens, independent of the user's choice. I don't know how to react on the button click directly. But you could register an "unload" event like this:
```py
unload = ui.button().on('unload', lambda: print('unload')).classes('hidden')
ui.add_body_html(f'''<script>
window.onunload = () => getElement({unload.id}).$emit('unload');
</script>
''')
```
